homeassistant/components/canary/camera.py

```diff
@@ -125,17 +128,36 @@ async def async_camera_image(
         self, width: int | None = None, height: int | None = None
     ) -> bytes | None:
         """Return a still image response from the camera."""
-        await self.hass.async_add_executor_job(self.renew_live_stream_session)
-        live_stream_url = await self.hass.async_add_executor_job(
-            getattr, self._live_stream_session, "live_stream_url"
-        )
-        return await ffmpeg.async_get_image(
-            self.hass,
-            live_stream_url,
-            extra_cmd=self._ffmpeg_arguments,
-            width=width,
-            height=height,
-        )
+        if self._image is None:
+            # if we still don't have an image, grab the live view
+            _LOGGER.debug("Grabbing a live view image from %s", self.name)
+            await self.hass.async_add_executor_job(self.renew_live_stream_session)
+
+            if self._live_stream_session is None:
+                return None
+
+            live_stream_url = self._live_stream_session.live_stream_url
+            if not live_stream_url:
+                return None
+
+            if self._image is None and live_stream_url:
+                image = await ffmpeg.async_get_image(
+                    self.hass,
+                    live_stream_url,
+                    extra_cmd=self._ffmpeg_arguments,
+                    width=width,
+                    height=height,
+                )
+
+                if image:
+                    self._image = image
+                    _LOGGER.debug("Grabbed a live view image from %s", self.name)
+                await self.hass.async_add_executor_job(
+                    self._live_stream_session.stop_session
+                )
+                _LOGGER.debug("Stopped live session from %s", self.name)
+
+        return self._image
```

Comment:
Shouldn't `self._image` be invalidated after some time?

Reply:
Good eye.. yes, the image will be invalidated/updated in the next PR that comes after this one [PR 73925](https://github.com/home-assistant/core/pull/73925).  This PR is 2/5 that were split out from 1 larger PR at @frenck's request.

Reply:
I had a similar thought as @emontnemery  -- Definitely makes sense to split into smaller PRs! but it still seems like there should be some kind of invalidation consideration in the same PR where the cache is added e.g. checking for the URL to change?  or invalidating after some expiration time.

I had a look at the other PRs and it's a little unclear about the order they should be reviewed given they have overlapping/duplicate changes.

Reply:
Thank you @allenporter, as I will have to go the multiple PR route again when adding in 2FA for canary after I finish testing that code. I don't want to create another mess at that time.  I will work on adding an expiration to the cache.

Reply:
marked the other pull requests as draft requests.. will enable when needed.